sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (C) 2020 chrysn
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @{
+ *
+ * @file
+ * @brief       CoAP file server implementation
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "checksum/fletcher32.h"
+#include "net/gcoap/fileserver.h"
+#include "net/gcoap.h"
+#include "vfs.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/** Maximum length of an expressible path, including the trailing 0 character. */
+#define COAPFILESERVER_PATH_MAX (64)
+
+/** Data extracted from a request on a file */
+struct requestdata {
+    /** 0-terminated expanded file name in the VFS */
+    char namebuf[COAPFILESERVER_PATH_MAX];
+    uint32_t blocknum2;
+    unsigned int szx2; /* would prefer uint8_t but that's what coap_get_blockopt gives */
+    uint32_t etag;
+    bool etag_sent;
+};
+
+/* These are almost but but not quite coap_handler_t -- they require the
+ * request to be already fully read and digested into the last argument. */
+static ssize_t coapfileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                                           struct requestdata *request);
+static ssize_t coapfileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
```

Comment:
Do you care to move the function definitions up here? This is just my personal taste. It can as well stay as it is.

Reply:
At least can't do them right there as per coding style:

> * Every function needs a prototype in addition to its definition. If a prototype
  is specified within a .c file it has to be declared BEFORE any function
  definitions.